shared/regex/codeql/regex/nfa/SuperlinearBackTracking.qll

```diff
@@ -264,41 +269,59 @@ module Make<RegexTreeViewSig TreeImpl> {
     result = [min(intersect(a, b)), max(intersect(a, b))]
   }
 
-  private newtype TTrace =
-    Nil() or
-    Step(InputSymbol s1, InputSymbol s2, InputSymbol s3, TTrace t) {
-      isReachableFromStartTuple(_, _, t, s1, s2, s3, _, _)
-    }
+  /** Gets a tuple reachable in a forwards exploratory search from the start state `(pivot, pivot, pivot)`. */
+  private StateTuple getReachableFromStartStateForwards(State pivot, State succ) {
+    // base case.
+    isStartLoops(pivot, succ) and
+    result = MkStateTuple(pivot, succ, pivot, pivot, succ)
+    or
+    // recursive case
+    exists(StateTuple p |
+      p = getReachableFromStartStateForwards(pivot, succ) and
+      step(p, _, _, _, result)
+    )
+  }
 
   /**
-   * A list of tuples of input symbols that describe a path in the product automaton
-   * starting from some start state.
+   * Gets a tuple tuple reachable from the end state `(succ, succ, succ)` in a backwards exploratory search.
+   * Where the end state was reachable from a forwards search from the start state `(pivot, pivot, pivot)`.
+   * The resulting tuples are exactly those that are on a path from the start state to the end state.
    */
-  class Trace extends TTrace {
-    /**
-     * Gets a string representation of this Trace that can be used for debug purposes.
-     */
-    string toString() {
-      this = Nil() and result = "Nil()"
-      or
-      exists(InputSymbol s1, InputSymbol s2, InputSymbol s3, Trace t | this = Step(s1, s2, s3, t) |
-        result = "Step(" + s1 + ", " + s2 + ", " + s3 + ", " + t + ")"
-      )
-    }
+  private StateTuple getARelevantStateTuple(State pivot, State succ) {
+    // base case.
+    isStartLoops(pivot, succ) and
+    result = MkStateTuple(pivot, succ, pivot, succ, succ) and
+    result = getReachableFromStartStateForwards(pivot, succ)
+    or
+    // recursive case
+    exists(StateTuple p |
+      p = getARelevantStateTuple(pivot, succ) and
+      step(result, _, _, _, p) and
+      pragma[only_bind_out](result) = getReachableFromStartStateForwards(pivot, succ) // was reachable in the forwards pass.
+    )
   }
 
   /**
    * Holds if there exists a transition from `r` to `q` in the product automaton.
+   * Where `r` and `q` are both on a path from a start state to an end state.
    * Notice that the arguments are flipped, and thus the direction is backwards.
    */
   pragma[noinline]
-  predicate tupleDeltaBackwards(StateTuple q, StateTuple r) { step(r, _, _, _, q) }
+  predicate tupleDeltaBackwards(StateTuple q, StateTuple r) {
```

Comment:
This might be less confusing if we rename the parameters to `dst` and `src` instead of `q` and `r`